src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,202 @@
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="value">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteFloatAsync(TdsStream stream, float value, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(!float.IsInfinity(value) && !float.IsNaN(value), "Float value is out of range.");
+
+            var len = sizeof(float);
+            BinaryPrimitives.WriteInt32LittleEndian(_alloc.GetBuffer(len), BitConverter.SingleToInt32Bits(value));
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes double value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="value">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteDoubleAsync(TdsStream stream, double value, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(!double.IsInfinity(value) && !double.IsNaN(value), "Double value is out of range.");
+
+            var len = sizeof(double);
+            BinaryPrimitives.WriteInt64LittleEndian(_alloc.GetBuffer(len), BitConverter.DoubleToInt64Bits(value));
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes partial long value to out buffer, as little-endian. This should be ideally used for writing date, 
+        /// time and related data types. The length specified should be between 0 to 8 bytes supported by long.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="value">Value to write</param>
+        /// <param name="length">Length to fill</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WritePartialLongAsync(TdsStream stream, long value, int length, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(length >= 0, "Length should not be negative");
+            Debug.Assert(length <= 8, "Length specified is longer than the size of a long");
+
+            var len = sizeof(long);
+            var buffer = _alloc.GetBuffer(len);
+            for (int i = 0; i < length; i++)
+            {
+                buffer[i] = (byte)((value >> (i * 8)) & 0xFF);
+            }
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, length, isAsync, ct).ConfigureAwait(false);
```

Comment:
Nit: I'd personally write the entire long to the buffer using `BinaryPrimitives`, then simply specify the length of the Span/Memory to be `length`. It'll get rid of the slightly awkward bit-shifting, eliminating a loop.

Reply:
This is specific to design requirements to write Date and Time and DateTime to Sql Server.